File: _stats_build/statistics/abstract/ao_rounds.rb
# NOTE: 抽象基类，用于计算跨轮次 average of averages 的 WR 历史
# AoXR = 一场比赛中某人恰好参加了 X 轮时，各轮 average 的均值
# 子类只需指定 round_count 即可
# 支持双视图 Tab：当前排名 + WR 历史
#
# HACK: 不走 Statistic 的统一 query → transform 流程。
# results 全表 550 万行，一次加载 ~8.7GB 导致 CI OOM。
# 改为按 event_id 逐个查询 MySQL，每次只加载 ~1-5 万行，
# 处理完后 GC 可回收，内存峰值 ~100MB。
#
# NOTE: 一次性计算模式——第一个子类运行时，为 4 种 round_count 同时计算结果。
# 后续子类直接取预计算数据，80 次 MySQL 查询减到 20 次。
require_relative "../../core/grouped_statistic"
require_relative "../../core/events"
require_relative "../../core/solve_time"
require_relative "../../core/tab_ui"
require_relative "../../core/database"

class AoRounds < GroupedStatistic
  include TabUi

  # NOTE: 预计算结果缓存
  # 结构：{ round_count => { history_data: [[event_name, wr_results], ...],
  #                          ranking_by_event: { event_name => [...] } } }
  @@precomputed = {}

  # NOTE: 所有子类的 round_count 枚举，用于一次性批量计算
  ROUND_COUNTS = [1, 2, 3, 4].freeze

  # NOTE: round_type_id 分类：用于确定各轮在 Details 中的排列顺序
  # 排序权重越小越靠前（R1 < R2 < R3 < Final）
  ROUND_SORT_ORDER = {
    '1' => 1, 'd' => 1,   # First round / Combined first
    '2' => 2, 'e' => 2,   # Second round / Combined second
    '3' => 3, 'g' => 3,   # Semi-final / Combined third
    'c' => 99, 'f' => 99  # Combined final / Final（总是最后）
  }.freeze

  def round_count
    raise "子类必须实现 round_count"
  end

  # NOTE: 按 event_id 单独查询，避免全表加载
  def query_for_event(event_id)
    <<-SQL
      SELECT
        result.event_id,
        result.average,
        result.round_type_id,
        result.competition_id,
        CONCAT('[', person.name, '](https://www.worldcubeassociation.org/persons/', person.wca_id, ')') person_link,
        person.wca_id AS person_id,
        CONCAT('[', competition.cell_name, '](https://www.worldcubeassociation.org/competitions/', competition.id, ')') competition_link,
        competition.start_date
      FROM results result
      JOIN persons person ON person.wca_id = result.person_id AND person.sub_id = 1
      JOIN competitions competition ON competition.id = result.competition_id
      WHERE result.average > 0 AND result.event_id = '#{event_id}'
      ORDER BY competition.start_date
    SQL
  end

  # NOTE: 重写 data，使用一次性计算模式
  # 第一个子类触发 compute_all_round_counts，后续子类直接取缓存
  def data
    return @data if @data

    compute_all_round_counts if @@precomputed.empty?

    result = @@precomputed[round_count]
    @ranking_by_event = result[:ranking_by_event]
    @data = result[:history_data]
  end

  # NOTE: 用 top + tabbed_grouped_markdown 替换原手写 HTML
  # 将 @ranking_by_event 中的 hash rows 归一化为数组供基类渲染
  def markdown
    ranking_header = { "Person" => :left, "Result" => :right, "Details" => :left }
    # NOTE: 必须先调 data 触发查询，才能填充 @ranking_by_event
    history_data = data
    # NOTE: build_ranking 返回 [{person_link:, metric_str:, details:}]，需转为数组格式
    ranking_data = @ranking_by_event.transform_values do |rows|
      rows.map { |r| [r[:person_link], r[:metric_str], r[:details]] }
    end
    top + tabbed_grouped_markdown(
      ranking_data: ranking_data,
      ranking_header: ranking_header,
      history_data: history_data,
      history_header: @table_header
    )
  end

  private

  # NOTE: 一次性为所有 round_count 计算结果
  # 每个 event 只查一次 MySQL，然后为 rc=1,2,3,4 分别计算
  # 处理完一个 event 后 raw rows 即可被 GC 回收
  def compute_all_round_counts
    events = Events::WITH_AVERAGE

    # 初始化结果容器
    ROUND_COUNTS.each do |rc|
      @@precomputed[rc] = { history_data: [], ranking_by_event: {} }
    end

    events.each do |event_id, event_name|
      # NOTE: 每个项目只查询 MySQL 一次，4 种 round_count 共享同一份数据
      event_rows = Database.client.query(query_for_event(event_id)).to_a

      # NOTE: 先做一次分组（最耗时的操作），4 种 rc 共用分组结果
      grouped = group_by_competition_person(event_rows, event_id)

      ROUND_COUNTS.each do |rc|
        computed = filter_and_compute(grouped, rc)
        @@precomputed[rc][:ranking_by_event][event_name] = build_ranking(computed, event_id)
        wr_results = build_wr_history(computed, event_id)
        @@precomputed[rc][:history_data] << [event_name, wr_results]
      end
      # event_rows 和 grouped 在 block 结束后可被 GC 回收
    end
  end

  # NOTE: 按 (competition_id, person_id) 分组，收集该人在该比赛的所有轮次记录
  def group_by_competition_person(event_rows, event_id)
    grouped = {}
    event_rows.each do |r|
      key = [r["competition_id"], r["person_id"]]
      grouped[key] ||= { "rows" => [], "person_link" => r["person_link"],
                         "person_id" => r["person_id"],
                         "competition_link" => r["competition_link"],
                         "start_date" => r["start_date"], "event_id" => event_id }
      grouped[key]["rows"] << r
    end
    grouped
  end

  # NOTE: 从已分组数据中筛选恰好有 rc 条记录的组，计算 AoXR
  # rc: round_count，由 compute_all_round_counts 传入
  def filter_and_compute(grouped, rc)
    grouped.filter_map do |_key, info|
      rows = info["rows"]
      next unless rows.size == rc

      # 按轮次排序（R1 < R2 < R3 < Final），提取 average 值
      sorted_rows = rows.sort_by { |r| ROUND_SORT_ORDER[r["round_type_id"]] || 50 }
      values = sorted_rows.map { |r| r["average"] }
      avg = values.sum.to_f / rc
      info.merge("_metric" => avg, "_round_values" => values)
    end
  end

  # NOTE: 每人最佳 metric，按项目 top 10
  def build_ranking(computed, event_id)
    best_by_person = {}
    computed.each do |r|
      pid = r["person_id"]
      if !best_by_person[pid] || r["_metric"] < best_by_person[pid]["_metric"]
        best_by_person[pid] = r
      end
    end

    best_by_person.values
      .sort_by { |r| r["_metric"] }
      .first(10)
      .map do |r|
        metric_str = SolveTime.new(event_id, :average, r["_metric"].round).clock_format
        details = r["_round_values"].map { |v| SolveTime.new(event_id, :average, v).clock_format }.join(', ')
        { person_link: r["person_link"], metric_str: metric_str, details: details }
      end
  end

  # NOTE: WR 历史——按日期正序扫描，只保留刷新最小值的记录，最终倒序
  def build_wr_history(computed, event_id)
    sorted = computed.sort_by { |r| r["start_date"] }
    min_so_far = Float::INFINITY
    wr_records = sorted.select do |r|
      if r["_metric"] < min_so_far
        min_so_far = r["_metric"]
        true
      else
        false
      end
    end

    results = wr_records.each_with_index.map do |r, i|
      metric_str = SolveTime.new(event_id, :average, r["_metric"].round).clock_format

      gain_str, days_str = wr_progress(wr_records, i) { |r| r["_metric"] }

      # Details: 显示各轮 average（已按轮次排序）
      details = r["_round_values"].map do |v|
        SolveTime.new(event_id, :average, v).clock_format
      end.join(', ')

      date_str = r["start_date"].strftime("%Y-%m-%d")
      [metric_str, gain_str, days_str, r["person_link"], r["competition_link"], date_str, details]
    end

    results.reverse
  end
end

File: core/packages/stats-build/src/core/ao_rounds.ts
// NOTE: AoRounds 抽象基类——跨轮次 average of averages
// AoXR = 一场比赛中某人恰好参加了 X 轮时，各轮 average 的均值
// 支持双视图 JSON：排名（ranking）+ WR 历史（history）
//
// NOTE: 一次性计算模式——第一个子类运行时，为 ROUND_COUNTS=[1,2,3,4]
// 四种 round_count 同时计算结果。每个 event 只查 MySQL 一次，
// 4 种 rc 共享同一份数据。
import { GroupedStatistic } from './grouped_statistic.js';
import { EVENTS_WITH_AVERAGE, headerZh, eventZh } from './events.js';
import { SolveTime } from './solve_time.js';
import { query as dbQuery } from './database.js';
import { formatDate, calcDays, filterWrHistory } from './format_date.js';
import type { StatJson, StatPanel, Alignment, TableHeader } from './statistic.js';
import type { RowDataPacket } from 'mysql2';

// NOTE: 排名表头
const RANKING_HEADER: TableHeader = {
  '#': 'right', 'Person': 'left', 'Result': 'right',
  'Country': 'left', 'Date': 'left', 'Competition': 'left', 'Details': 'left',
};

// NOTE: round_type_id 排序权重
const ROUND_SORT_ORDER: Record<string, number> = {
  '1': 1, 'd': 1, '2': 2, 'e': 2, '3': 3, 'g': 3, 'c': 99, 'f': 99,
};

// NOTE: 所有子类的 round_count 枚举——一次性为 4 种 rc 批量计算
const ROUND_COUNTS = [1, 2, 3, 4] as const;

// --- 预计算缓存 ---
// 结构：roundCount -> { historyData, rankingData }
interface AoRoundsCache {
  historyData: [string, unknown[][]][];
  rankingData: [string, unknown[][]][];
}
let precomputed: Map<number, AoRoundsCache> | null = null;

export abstract class AoRounds extends GroupedStatistic {
  // NOTE: 子类必须实现
  abstract roundCount(): number;

  constructor() {
    super();
    this.tableHeader = {
      'Result': 'right', 'Improvement': 'right', 'Days': 'right',
      'Person': 'left', 'Date': 'left', 'Competition': 'left', 'Details': 'left',
    };
  }

  // NOTE: 按 event_id 单独查询，避免全表加载
  private queryForEvent(eventId: string): string {
    return `
      SELECT
        result.event_id,
        result.average,
        result.round_type_id,
        result.competition_id,
        CONCAT('[', person.name, '](https://www.worldcubeassociation.org/persons/', person.wca_id, ')') person_link,
        person.wca_id AS person_id,
        person.country_id,
        CONCAT('[', competition.cell_name, '](https://www.worldcubeassociation.org/competitions/', competition.id, ')') competition_link,
        competition.start_date
      FROM results result
      JOIN persons person ON person.wca_id = result.person_id AND person.sub_id = 1
      JOIN competitions competition ON competition.id = result.competition_id
      WHERE result.average > 0 AND result.event_id = '${eventId}'
      ORDER BY competition.start_date
    `;
  }

  query(): string { return ''; }  // NOTE: 不使用统一 query，逐 event 查询

  // NOTE: toJson() 覆写了整个流程，transform 不被调用，提供空实现满足抽象类
  transform(): [string, unknown[][]][] { return []; }

  // NOTE: 一次性为所有 round_count 预计算结果
  // 每个 event 只查 MySQL 一次，4 种 rc 共享同一份数据
  static async precomputeAllRoundCounts(): Promise<void> {
    precomputed = new Map();

    // NOTE: 初始化 4 种 rc 的结果容器
    for (const rc of ROUND_COUNTS) {
      precomputed.set(rc, { historyData: [], rankingData: [] });
    }

    const events = EVENTS_WITH_AVERAGE;

    for (const [eventId, eventName] of Object.entries(events)) {
      // NOTE: 每个项目只查 MySQL 一次，4 种 rc 共享
      const eventRows = await dbQuery<RowDataPacket[]>(`
        SELECT
          result.event_id,
          result.average,
          result.round_type_id,
          result.competition_id,
          CONCAT('[', person.name, '](https://www.worldcubeassociation.org/persons/', person.wca_id, ')') person_link,
          person.wca_id AS person_id,
          person.country_id,
          CONCAT('[', competition.cell_name, '](https://www.worldcubeassociation.org/competitions/', competition.id, ')') competition_link,
          competition.start_date
        FROM results result
        JOIN persons person ON person.wca_id = result.person_id AND person.sub_id = 1
        JOIN competitions competition ON competition.id = result.competition_id
        WHERE result.average > 0 AND result.event_id = '${eventId}'
        ORDER BY competition.start_date
      `);

      // NOTE: 先做一次分组（最耗时），4 种 rc 共用
      const grouped = new Map<string, { rows: RowDataPacket[]; meta: RowDataPacket }>();
      for (const r of eventRows) {
        const key = `${r['competition_id']}:${r['person_id']}`;
        const existing = grouped.get(key);
        if (existing) {
          existing.rows.push(r);
        } else {
          grouped.set(key, { rows: [r], meta: r });
        }
      }

      // NOTE: 对每种 rc 筛选、计算排名和 WR 历史
      for (const rc of ROUND_COUNTS) {
        const computed = filterAndCompute(grouped, rc);
        const ranking = buildRanking(computed, eventId);
        const history = buildWrHistory(computed, eventId);
        precomputed.get(rc)!.rankingData.push([eventName, ranking]);
        precomputed.get(rc)!.historyData.push([eventName, history]);
      }
      // NOTE: eventRows 和 grouped 在此 block 结束后可被 GC 回收
    }
  }

  // NOTE: 缓存清理——wr_aoxr 聚合完成后调用
  static clearPrecomputed(): void {
    precomputed = null;
  }

  // NOTE: 覆写 toJson——双视图 panels 输出
  async toJson(): Promise<StatJson> {
    const rc = this.roundCount();

    // NOTE: 如果缓存为空，一次性为所有 round_count 预计算
    if (!precomputed) {
      await AoRounds.precomputeAllRoundCounts();
    }

    const cached = precomputed!.get(rc);
    let historyData: [string, unknown[][]][];
    let rankingData: [string, unknown[][]][];

    if (cached) {
      // NOTE: 缓存命中——直接使用预计算结果
      historyData = cached.historyData;
      rankingData = cached.rankingData;
    } else {
      // NOTE: 缓存未命中（极端情况）——回退到独立计算
      const result = await this.computeIndependently();
      historyData = result.historyData;
      rankingData = result.rankingData;
    }

    // NOTE: 构建 panels
    const historyHeaderEntries = Object.entries(this.tableHeader);
    const rankingHeaderEntries = Object.entries(RANKING_HEADER);

    const buildSections = (data: [string, unknown[][]][]) =>
      data.filter(([, rows]) => rows.length > 0)
        .map(([title, rows]) => ({ title, titleZh: eventZh(title), rows }));

    const panels: StatPanel[] = [
      {
        id: 'ranking', labelEn: 'Ranking', labelZh: '排名',
        header: rankingHeaderEntries.map(([label, align]) => ({
          key: label.toLowerCase().replace(/\s+/g, '_'),
          label, labelZh: headerZh(label), align,
        })),
        sections: buildSections(rankingData),
      },
      {
        id: 'history', labelEn: 'History', labelZh: '历史',
        header: historyHeaderEntries.map(([label, align]) => ({
          key: label.toLowerCase().replace(/\s+/g, '_'),
          label, labelZh: headerZh(label), align,
        })),
        sections: buildSections(historyData),
      },
    ];

    return {
      id: this.id,
      title: this.title,
      titleZh: this.titleZh || this.title,
      ...(this.note ? { note: this.note } : {}),
      ...(this.noteZh ? { noteZh: this.noteZh } : {}),
      header: [],
      panels,
    };
  }

  // NOTE: 回退独立计算——仅在缓存未命中时使用
  private async computeIndependently(): Promise<AoRoundsCache> {
    const events = EVENTS_WITH_AVERAGE;
    const rc = this.roundCount();
    const historyData: [string, unknown[][]][] = [];
    const rankingData: [string, unknown[][]][] = [];

    for (const [eventId, eventName] of Object.entries(events)) {
      const eventRows = await dbQuery<RowDataPacket[]>(this.queryForEvent(eventId));

      const grouped = new Map<string, { rows: RowDataPacket[]; meta: RowDataPacket }>();
      for (const r of eventRows) {
        const key = `${r['competition_id']}:${r['person_id']}`;
        const existing = grouped.get(key);
        if (existing) {
          existing.rows.push(r);
        } else {
          grouped.set(key, { rows: [r], meta: r });
        }
      }

      const computed = filterAndCompute(grouped, rc);
      const ranking = buildRanking(computed, eventId);
      const history = buildWrHistory(computed, eventId);
      rankingData.push([eventName, ranking]);
      historyData.push([eventName, history]);
    }

    return { historyData, rankingData };
  }
}

// --- 辅助函数 ---

interface ComputedEntry {
  metric: number;
  roundValues: number[];
  meta: RowDataPacket;
}

// NOTE: 从已分组数据中筛选恰好有 rc 条记录的组，计算 AoXR
function filterAndCompute(
  grouped: Map<string, { rows: RowDataPacket[]; meta: RowDataPacket }>,
  rc: number,
): ComputedEntry[] {
  const result: ComputedEntry[] = [];
  for (const [, info] of grouped) {
    if (info.rows.length !== rc) continue;
    const sorted = info.rows.sort((a, b) =>
      (ROUND_SORT_ORDER[String(a['round_type_id'])] ?? 50)
      - (ROUND_SORT_ORDER[String(b['round_type_id'])] ?? 50));
    const values = sorted.map(r => Number(r['average']));
    const avg = values.reduce((s, v) => s + v, 0) / rc;
    result.push({ metric: avg, roundValues: values, meta: info.meta });
  }
  return result;
}

// NOTE: 每人最佳 metric，按项目排 top 10
function buildRanking(computed: ComputedEntry[], eventId: string): unknown[][] {
  const bestByPerson = new Map<string, ComputedEntry>();
  for (const c of computed) {
    const pid = String(c.meta['person_id']);
    const existing = bestByPerson.get(pid);
    if (!existing || c.metric < existing.metric) {
      bestByPerson.set(pid, c);
    }
  }

  return [...bestByPerson.values()]
    .sort((a, b) => a.metric - b.metric)
    .slice(0, 10)
    .map((v, i) => {
      const metricStr = new SolveTime(eventId, 'average', Math.round(v.metric)).clockFormat();
      const detailsCsv = v.roundValues
        .map(val => new SolveTime(eventId, 'average', val).clockFormat())
        .join(',');
      const dateStr = formatDate(v.meta['start_date']);
      return [i + 1, v.meta['person_link'], metricStr, v.meta['country_id'], dateStr, v.meta['competition_link'], { _type: 'solves' as const, csv: detailsCsv }];
    });
}

// NOTE: WR 历史——按日期正序扫描，只保留刷新最小值的记录
function buildWrHistory(computed: ComputedEntry[], eventId: string): unknown[][] {
  // NOTE: filterWrHistory 内置 formatDate 日期排序（YYYY-MM-DD）+ 同日期大值先扫描 + <= minSoFar
  // strict: true 保持原来的 < minSoFar 语义（AoXR 不含平 WR）
  const wrRecords = filterWrHistory(
    computed,
    c => c.meta['start_date'],
    c => c.metric,
    { strict: true },
  );

  const historyRows = wrRecords.map((c, i) => {
    const metricStr = new SolveTime(eventId, 'average', Math.round(c.metric)).clockFormat();
    const detailsCsv = c.roundValues
      .map(v => new SolveTime(eventId, 'average', v).clockFormat())
      .join(',');

    // NOTE: 进步百分比
    let gainStr = '';
    if (i > 0) {
      const prev = wrRecords[i - 1].metric;
      gainStr = `${((prev - c.metric) / prev * 100).toFixed(1)}%`;
    }

    // NOTE: 天数——用共享 calcDays 工具
    const nextDateVal = i < wrRecords.length - 1 ? wrRecords[i + 1].meta['start_date'] : null;
    const daysStr = calcDays(c.meta['start_date'], nextDateVal);

    const dateStr = formatDate(c.meta['start_date']);
    return [metricStr, gainStr, daysStr, c.meta['person_link'], dateStr, c.meta['competition_link'], { _type: 'solves' as const, csv: detailsCsv }];
  });

  return historyRows.reverse();
}
